lib/innit-client/tests/integration_test/mod.rs

```diff
@@ -0,0 +1,111 @@
+mod certs;
+#[cfg(test)]
+mod integration_tests {
+
+    use base64::{Engine as _, engine::general_purpose::STANDARD};
+    use innit_client::InnitClientError;
+    use std::net::SocketAddr;
+
+    use hyper::server::conn::AddrIncoming;
+    use innit_client::config::StandardConfig;
+    use innit_client::{InnitClient, auth::AuthConfig};
+    use innit_server::Server;
+    use si_tls::{CertificateSource, KeySource};
+    use tokio_util::sync::CancellationToken;
+    use url::Url;
+
+    use super::certs::*;
+
+    struct TestCerts {
+        ca_cert: String,
+        client_cert: String,
+        client_key: String,
+    }
+
+    fn create_certs() -> TestCerts {
+        rustls::crypto::ring::default_provider()
+            .install_default()
+            .expect("Failed to install rustls crypto provider");
+        let (ca, ca_key) = new_ca();
+        let (client_cert, client_key) = new_end_entity(&ca, &ca_key);
+
+        TestCerts {
+            ca_cert: STANDARD.encode(ca.pem()),
+            client_cert: STANDARD.encode(client_cert.pem()),
+            client_key: STANDARD.encode(client_key.serialize_pem()),
+        }
+    }
+
+    async fn setup_test_server(
+        addr: SocketAddr,
+        ca_cert: String,
+    ) -> (Server<AddrIncoming>, CancellationToken) {
+        let token = CancellationToken::new();
+        let config = innit_server::Config::builder()
+            .socket_addr(addr)
+            .client_ca_cert(Some(CertificateSource::Base64(ca_cert)))
+            .build()
+            .expect("should build config");
+        let server = Server::http(config, token.clone())
+            .await
+            .expect("should create server");
+        (server, token)
+    }
+
+    #[tokio::test]
+    async fn test_client_server_mtls_interaction() {
+        let certs = create_certs();
+        let addr = SocketAddr::from(([0, 0, 0, 0], 5166));
```

Comment:
I *believe* you can pass a port of `0` in here which tells the kernel to assign a random/unallocated port for this socket. Then you can ask for that port number via [`addr.port()`](https://doc.rust-lang.org/std/net/enum.SocketAddr.html#method.port) which should allow for better test concurrency in the future.

Ok, I'm pretty sure this works, according to [users.rust-lang.org](https://users.rust-lang.org/t/what-does-the-port-0-represent/122583/4)